lib/handler/debug_state.c

```diff
@@ -0,0 +1,236 @@
+        }                                                                                        \
+    } while (0)
+
+    // stringify these variables to use later in Debug Header
+    h2o_iovec_t conn_flow_out;
+    conn_flow_out.base = h2o_mem_alloc_pool(&req->pool, sizeof(H2O_INT64_LONGEST_STR));
+    conn_flow_out.len = sprintf(conn_flow_out.base, "%zd", conn->_write.window._avail);
+
+
+    h2o_iovec_t conn_flow_in;
+    conn_flow_in.base = h2o_mem_alloc_pool(&req->pool, sizeof(H2O_INT64_LONGEST_STR));
+    conn_flow_in.len = sprintf(conn_flow_in.base, "%zd", conn->_input_window._avail);
+
+    OUTPUT("{\n"
+         "  \"settings\": {\n"
+         "    \"SETTINGS_HEADER_TABLE_SIZE\": %" PRIu32 ",\n"
+         "    \"SETTINGS_ENABLE_PUSH\": %" PRIu32 ",\n"
+         "    \"SETTINGS_MAX_CONCURRENT_STREAMS\": %" PRIu32 ",\n"
+         "    \"SETTINGS_INITIAL_WINDOW_SIZE\": %" PRIu32 ",\n"
+         "    \"SETTINGS_MAX_FRAME_SIZE\": %" PRIu32 "\n"
+         "  },\n"
+         "  \"peerSettings\": {\n"
+         "    \"SETTINGS_HEADER_TABLE_SIZE\": %" PRIu32 ",\n"
+         "    \"SETTINGS_ENABLE_PUSH\": %" PRIu32 ",\n"
+         "    \"SETTINGS_MAX_CONCURRENT_STREAMS\": %" PRIu32 ",\n"
+         "    \"SETTINGS_INITIAL_WINDOW_SIZE\": %" PRIu32 ",\n"
+         "    \"SETTINGS_MAX_FRAME_SIZE\": %" PRIu32 "\n"
+         "  },\n"
+         "  \"connFlowOut\": %s,\n"
+         "  \"connFlowIn\": %s,\n"
+         "  \"streams\": {",
+         H2O_HTTP2_SETTINGS_HOST.header_table_size,
+         H2O_HTTP2_SETTINGS_HOST.enable_push,
+         H2O_HTTP2_SETTINGS_HOST.max_concurrent_streams,
+         H2O_HTTP2_SETTINGS_HOST.initial_window_size,
+         H2O_HTTP2_SETTINGS_HOST.max_frame_size,
+         conn->peer_settings.header_table_size,
+         conn->peer_settings.enable_push,
+         conn->peer_settings.max_concurrent_streams,
+         conn->peer_settings.initial_window_size,
+         conn->peer_settings.max_frame_size,
+         conn_flow_out.base,
+         conn_flow_in.base);
+
+    /* encode streams */
+    {
+        int coma_removed = 0;
+        h2o_http2_stream_t *stream;
+        kh_foreach_value(conn->streams, stream, {
+            const char *state_string = get_debug_state_string(stream);
+            if (state_string == NULL)
+                continue;
+
+            OUTPUT(",\n"
+                   "    \"%" PRIu32 "\": {\n"
+                   "      \"state\": \"%s\",\n"
+                   "      \"flowIn\": %zd,\n"
+                   "      \"flowOut\": %zd,\n"
+                   "      \"dataIn\": %zu,\n"
+                   "      \"dataOut\": %zu\n"
+                   "    }",
+                   stream->stream_id,
+                   state_string,
+                   stream->input_window._avail,
+                   stream->output_window._avail,
+                   (stream->_req_body == NULL ? 0 : stream->_req_body->size),
+                   stream->req.bytes_sent);
```

Comment:
I would appreciate it if we could use memory less aggressively. In the current implementation, `OUTPUT` allocates 2KB chunk, and that means that we would consume at max 2*(100+100)=400KB of memory here.